src/typesystem.cpp

```diff
@@ -37,7 +37,14 @@ MonoId TypeSystemCore::new_var() {
 
 MonoId TypeSystemCore::new_term(
     TypeFunctionId type_function, std::vector<int> args, char const* tag) {
-	int argument_count = type_function_data[type_function].argument_count;
+	TypeFunctionId tf_id = func_find(type_function);
+	TypeFunctionData& tf_data = type_function_data[tf_id];
+
+	// TODO: fail on dummy type function?
+	if (tf_data.type == TypeFunctionTag::Var)
+		assert(0 && "instantiating type function that could not be deduced");
```

Comment:
Oh I hadn't noticed but, this check is wrong.

We should be able to create a term with a typefunc that isnt known. We just need to go back at the end and verify that we do eventually learn what it is.

Reply:
Of course, in those cases we can't check that the argument is right. That also has to be done after type checking.

Reply:
The same thing goes for dummies. It should be possible to instanciate one, but we should make sure that it eventually gets resolved.